core/src/main/java/org/elasticsearch/index/engine/Engine.java

```diff
@@ -277,9 +278,119 @@ public Condition newCondition() {
         }
     }
 
-    public abstract void index(Index operation) throws EngineException;
+    public abstract IndexResult index(Index operation);
 
-    public abstract void delete(Delete delete) throws EngineException;
+    public abstract DeleteResult delete(Delete delete);
+
+    /**
+     * Base class for index and delete operation results
+     * Holds result meta data (e.g. translog location, updated version)
+     * for an executed write {@link Operation}
+     **/
+    public abstract static class Result {
+        private final Operation.TYPE operationType;
+        private final long version;
+        private final Exception failure;
+        private Translog.Location translogLocation;
+        private long took;
+        private boolean freeze;
```

Comment:
maybe use `private final SetOnce<Boolean> frozen` here it will barf if you freeze multiple times?